From 0a34b0b1da2e0042553d2cc6eda5cf49dfaafa71 Mon Sep 17 00:00:00 2001
From: Juliusz Sosinowicz <juliusz@wolfssl.com>
Date: Fri, 4 Sep 2026 09:41:51 +0000
Subject: [PATCH] wolfSSL Patch

This patch adds a --with-wolfssl=DIR configure option that builds tcpdump
against wolfSSL instead of OpenSSL libcrypto. The wolfSSL headers are
picked up through their OpenSSL compatibility layer, so no source changes
are needed beyond the configure script. DEBUG_WOLFSSL builds turn on
wolfSSL logging at startup.

Compile wolfSSL with:
  ./configure --enable-tcpdump
  make
  make install

--enable-tcpdump turns on OPENSSL_EXTRA, DES3 (ESP 3DES-CBC decryption)
and MD5 (TCP MD5 signature option). Add --enable-debug for wolfSSL logging.

Compile tcpdump with:
  patch -p1 < tcpdump-4.99.6.patch
  autoreconf -ivf
  ./configure --with-wolfssl=/usr/local
  make

Adjust /usr/local to the wolfSSL install prefix. When the prefix is not on
the dynamic linker search path, export LD_LIBRARY_PATH=<prefix>/lib before
running tcpdump or the tests.

Run the tests with:
  make check

All tests pass (636 passed, 0 failed, 11 skipped). The skipped tests are
the same as with an OpenSSL build: bgp-as-path-oobr-nossl and
bgp-aigp-oobr-nossl (HAVE_LIBCRYPTO is set), the smb_*/eapon1*/ipx/
nbns-valgrind tests (ENABLE_SMB not set) and lmp-v-fptype2.

libpcap 1.10.4 (as shipped by Ubuntu 24.04) reads pcap timestamps as
signed 32-bit values, so time_2038_overflow, time_2039, time_2106 and
time_2106_max fail with it regardless of the crypto library. Build libpcap
1.10.5 or newer in ../libpcap (tcpdump's configure picks it up on its own)
or install it system-wide.
---
 configure.ac | 51 ++++++++++++++++++++++++++++++++++++++++++++++++---
 tcpdump.c    |  7 +++++++
 2 files changed, 55 insertions(+), 3 deletions(-)

diff --git a/configure.ac b/configure.ac
index b50fef3..192bfaf 100644
--- a/configure.ac
+++ b/configure.ac
@@ -998,6 +998,40 @@ AC_CHECK_TOOL([AR], [ar])
 
 AC_LBL_DEVEL(V_CCOPT)
 
+AC_ARG_WITH(wolfssl,
+    AS_HELP_STRING([--with-wolfssl]@<:@=DIR@:>@,
+		   [use wolfSSL (located in directory DIR) @<:@default=no@:>@]),
+[
+	if test "x$withval" = "xno"
+	then
+		USE_WOLFSSL=no
+	elif test "x$withval" = "xyes"
+	then
+		AC_MSG_ERROR([--with-wolfssl needs the wolfSSL install directory: --with-wolfssl=DIR])
+	else
+		CPPFLAGS="$CPPFLAGS -I${withval}/include/wolfssl -I${withval}/include/ -DWOLFSSL_USE_OPTIONS_H"
+		LDFLAGS="$LDFLAGS -L${withval}/lib"
+		LIBS="$LIBS -lwolfssl"
+		USE_WOLFSSL=yes
+	fi],
+[
+	USE_WOLFSSL=no
+])
+
+if test "x$USE_WOLFSSL" = "xyes"
+then
+    AC_CHECK_HEADER([wolfssl/options.h])
+    if test $ac_cv_header_wolfssl_options_h = yes
+    then
+        AC_DEFINE(HAVE_WOLFSSL, 1, [define if you're using wolfSSL])
+        AC_DEFINE(HAVE_EVP_CIPHER_CTX_NEW, 1, [])
+        AC_DEFINE(HAVE_EVP_DECRYPTINIT_EX, 1, [])
+        AC_DEFINE(HAVE_LIBCRYPTO, 1, [])
+    else
+        AC_MSG_ERROR([Unable to find wolfssl.])
+    fi
+fi
+
 # Check for OpenSSL/libressl libcrypto
 AC_MSG_CHECKING(whether to use OpenSSL/libressl libcrypto)
 # Specify location for both includes and libraries.
@@ -1006,7 +1040,12 @@ AC_ARG_WITH(crypto,
     AS_HELP_STRING([--with-crypto]@<:@=DIR@:>@,
 		   [use OpenSSL/libressl libcrypto (located in directory DIR, if specified) @<:@default=yes, if available@:>@]),
 [
-	if test $withval = no
+	if test "x$USE_WOLFSSL" = "xyes"
+	then
+		want_libcrypto=no
+		AC_MSG_RESULT(no)
+		AC_MSG_WARN([--with-crypto is ignored when --with-wolfssl is given])
+	elif test $withval = no
 	then
 		# User doesn't want to link with libcrypto.
 		want_libcrypto=no
@@ -1025,12 +1064,18 @@ AC_ARG_WITH(crypto,
 		AC_MSG_RESULT([yes, using the version installed in $withval])
 	fi
 ],[
+	if test "x$USE_WOLFSSL" = "xyes"
+	then
+		want_libcrypto=no
+		AC_MSG_RESULT(no)
 	#
 	# Use libcrypto if it's present, otherwise don't; no directory
 	# was specified.
 	#
-	want_libcrypto=ifavailable
-	AC_MSG_RESULT([yes, if available])
+	else
+		want_libcrypto=ifavailable
+		AC_MSG_RESULT([yes, if available])
+	fi
 ])
 if test "$want_libcrypto" != "no"; then
 	#
diff --git a/tcpdump.c b/tcpdump.c
index bba632c..8d78591 100644
--- a/tcpdump.c
+++ b/tcpdump.c
@@ -69,6 +69,9 @@ The Regents of the University of California.  All rights reserved.\n";
 #ifdef HAVE_LIBCRYPTO
 #include <openssl/crypto.h>
 #endif
+#if defined(HAVE_WOLFSSL) && defined(DEBUG_WOLFSSL)
+#include <wolfssl/wolfcrypt/logging.h>
+#endif
 
 #ifdef HAVE_GETOPT_LONG
 #include <getopt.h>
@@ -1713,6 +1716,10 @@ main(int argc, char **argv)
 	memset(ndo, 0, sizeof(*ndo));
 	ndo_set_function_pointers(ndo);
 
+#if defined(HAVE_WOLFSSL) && defined(DEBUG_WOLFSSL)
+	wolfSSL_Debugging_ON();
+#endif
+
 	cnt = -1;
 	device = NULL;
 	infile = NULL;
-- 
2.43.0

